clippy_lints/src/avoidable_slice_indexing.rs

```diff
@@ -0,0 +1,291 @@
+        }
+    }
+}
+
+fn filter_lintable_slices(
+    cx: &LateContext<'_>,
+    slice_lint_info: FxHashMap<hir::HirId, SliceLintInformation>,
+    index_limit: u64,
+    scope: &hir::Expr<'_>,
+) -> Option<FxHashMap<hir::HirId, SliceLintInformation>> {
+    let mut v = SliceIndexLintingDelegate {
+        cx,
+        slice_lint_info,
+        index_limit,
+    };
+
+    cx.tcx.infer_ctxt().enter(|infcx| {
+        ExprUseVisitor::new(&mut v, &infcx, scope.hir_id.owner, cx.param_env, cx.typeck_results()).consume_expr(scope);
+    });
+
+    let slices: FxHashMap<hir::HirId, SliceLintInformation> = v
+        .slice_lint_info
+        .drain()
+        .filter(|(_key, value)| !value.index_use.is_empty())
+        .collect();
+    (!slices.is_empty()).then(|| slices)
+}
+
+struct SliceIndexLintingDelegate<'a, 'tcx> {
+    cx: &'a LateContext<'tcx>,
+    slice_lint_info: FxHashMap<hir::HirId, SliceLintInformation>,
+    index_limit: u64,
+}
+
+impl<'a, 'tcx> SliceIndexLintingDelegate<'a, 'tcx> {
+    fn check_place(&mut self, place: &PlaceWithHirId<'tcx>, hir_id: hir::HirId) {
+        if let PlaceBase::Local(slice_id) = place.place.base {
+            if !self.slice_lint_info.contains_key(&slice_id) {
+                return;
+            }
+
+            // Get the parent expression
+            let map = self.cx.tcx.hir();
+            let parent_id = map.get_parent_node(hir_id);
+            let parent_expr = if let Some(hir::Node::Expr(parent_expr)) = map.find(parent_id) {
+                parent_expr
+            } else {
+                return;
+            };
+
+            // Checking for slice indexing
+            if_chain! {
+                if let hir::ExprKind::AddrOf(_kind, _mut, inner_expr) = parent_expr.kind;
+                if let hir::ExprKind::Index(_, index_expr) = inner_expr.kind;
+                if let Some((Constant::Int(index_value), _)) = constant(self.cx, self.cx.typeck_results(), index_expr);
+                if let Ok(index_value) = index_value.try_into();
+                if index_value < self.index_limit;
+                if let Some(use_info) = self.slice_lint_info.get_mut(&slice_id);
+                then {
+                    use_info.index_use.push((index_value, map.span(parent_expr.hir_id)));
+                    return;
+                }
+            }
+
+            // The slice was used for something else than indexing
+            self.slice_lint_info.remove(&slice_id);
+        }
+    }
+}
+
+impl<'a, 'tcx> Delegate<'tcx> for SliceIndexLintingDelegate<'a, 'tcx> {
+    fn consume(&mut self, place: &PlaceWithHirId<'tcx>, expr_id: hir::HirId) {
+        self.check_place(place, expr_id);
+    }
+
+    fn borrow(&mut self, place: &PlaceWithHirId<'tcx>, expr_id: hir::HirId, _bk: ty::BorrowKind) {
+        self.check_place(place, expr_id);
+    }
+
+    fn mutate(&mut self, place: &PlaceWithHirId<'tcx>, expr_id: hir::HirId) {
+        self.check_place(place, expr_id);
+    }
```

Comment:
It seems like this could be accomplished with a normal visitor and `path_to_local_id`?

Reply:
Also true, I at first used the provided information by the `ExprUseVisitor` but then switched over to this. Nice catch